### PR TITLE
[unsafe-fields] Add as_ref_checked

### PR DESCRIPTION
**Pull request by joshlf**
Release 0.2.0.

Makes progress on #1931

---

This PR is on branch [unsafe-fields-zerocopy](../tree/unsafe-fields-zerocopy).

- #1960

### CODE REVIEW

File: unsafe-fields/src/lib.rs [hunk 1]
```diff
@@ -220,6 +225,26 @@ impl<O: ?Sized, F: ?Sized, const NAME_HASH: u128> Unsafe<O, F, { NAME_HASH }> {
...
+    /// immutable.
+    ///
+    /// If [`F: Immutable`][immutable], then `F` does not permit interior
+    /// mutation, and so it is safe to return a reference to it.
+    ///
+    /// [immutable]: zerocopy_0_8::Immutable
+    #[inline(always)]
+    #[cfg(feature = "zerocopy_0_8")]
+    #[cfg_attr(doc_cfg, doc(cfg(feature = "zerocopy_0_8")))]
+    pub const fn as_ref_checked(&self) -> &F
```
Comment (jswrenn) [inline unsafe-fields/src/lib.rs:238]: My preference would be `as_ref_unchecked` for the unsafe method, and make the safe thing the shortest to type (`.as_ref`). Most types are `Immutable`, so I additionally expect the safe version will be the common case!

Comment (joshlf) [reply unsafe-fields/src/lib.rs]: Feels weird to have most of the API suffixed with `_unchecked`, but I agree that that's philosophically the right thing to do.

Done.

File: unsafe-fields/src/lib.rs [hunk 2]
```diff
@@ -215,33 +219,53 @@ impl<O: ?Sized, F: ?Sized, const NAME_HASH: u128> Unsafe<O, F, { NAME_HASH }> {
...
+
     /// Gets a mutable reference to the inner value.
     ///
     /// # Safety
     ///
     /// The caller is responsible for upholding any safety invariants associated
     /// with this field.
     #[inline(always)]
-    pub unsafe fn as_mut(&mut self) -> &mut F {
+    pub unsafe fn as_mut_unchecked(&mut self) -> &mut F {
```
Comment (jswrenn) [inline unsafe-fields/src/lib.rs:254]: Same concern as `new`. I don't *think* there's a safe version we could possibly offer here that we'd want to distinguish the unsafe version from.

**Review by jswrenn — COMMENTED** [review]
I don't think we necessarily need to name *everything* `_unchecked`. The `_unchecked` prefix in Rust is conventionally used in Rust when there's a safe option that the method needs to be distinguished from.

When there's no safe alternative, the standard library omits the `_unchecked` suffix.

File: unsafe-fields/src/lib.rs (hunk 2 above)
Comment (joshlf) [reply unsafe-fields/src/lib.rs:254]: Done.

File: unsafe-fields/src/lib.rs [hunk 3]
```diff
@@ -215,33 +219,53 @@ impl<O: ?Sized, F: ?Sized, const NAME_HASH: u128> Unsafe<O, F, { NAME_HASH }> {
...
-    /// Constructs a new `Unsafe<O, F, NAME_HASH>`.
+    /// Constructs a new `Unsafe`.
     ///
     /// # Safety
     ///
     /// The caller is responsible for upholding any safety invariants associated
     /// with this field.
     #[inline(always)]
-    pub const unsafe fn new(field: F) -> Unsafe<O, F, { NAME_HASH }> {
+    pub const unsafe fn new_unchecked(field: F) -> Unsafe<O, F, { NAME_HASH }> {
```
[the remark this answers is not in the document]
Comment (joshlf) [reply unsafe-fields/src/lib.rs]: Done.

File: unsafe-fields/src/lib.rs [hunk 4]
```diff
@@ -253,7 +277,7 @@ impl<O: ?Sized, F, const NAME_HASH: u128> Unsafe<O, F, { NAME_HASH }> {
     /// The caller is responsible for upholding any safety invariants associated
     /// with this field.
     #[inline(always)]
-    pub const unsafe fn into(self) -> F {
+    pub const unsafe fn into_unchecked(self) -> F {
```
[the remark this answers is not in the document]
Comment (joshlf) [reply unsafe-fields/src/lib.rs]: Done.

**Review by jswrenn — APPROVED** [review]

**Pull request merged**